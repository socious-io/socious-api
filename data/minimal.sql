CREATE TABLE public.users(
  id serial PRIMARY KEY,
  first_name varchar(70) DEFAULT NULL,
  username varchar(200) UNIQUE NOT NULL,
  name varchar(255) DEFAULT NULL,
  email varchar(200) UNIQUE NOT NULL,
  email_text varchar(255) DEFAULT NULL,
  mobile_countries_id int(11) DEFAULT NULL,
  phone varchar(255) DEFAULT NULL,
  wallet_address varchar(255) DEFAULT NULL,
  password text NOT NULL,
  remember_token text DEFAULT NULL,
  city text,
  city_ja mediumtext,
  description_search mediumtext,
  description_search_ja mediumtext,
  address_detail text,
  avatar text,
  cover_image text,
  expiry_date datetime DEFAULT NULL,
  active boo,
  activation_token varchar(255) DEFAULT NULL,
  mission text,
  bio mediumtext,
  profile_id varchar(200) DEFAULT NULL,
  view_as int(11) DEFAULT NULL,
  language varchar(255) DEFAULT NULL,
  my_conversation varchar(255) DEFAULT NULL,  
  email_verified_at datetime DEFAULT NULL,
  impact_score float NOT NULL DEFAULT '0',
  lock_otp datetime DEFAULT NULL,
  created_at datetime NOT NULL DEFAULT CURRENT_TIMESTAMP(6),
  updated_at datetime NOT NULL DEFAULT CURRENT_TIMESTAMP(6) ON UPDATE CURRENT_TIMESTAMP(6),
  deleted_at datetime DEFAULT NULL
);
