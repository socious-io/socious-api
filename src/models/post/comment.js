import sql from 'sql-template-tag';
import {app} from '../../index.js';
import {EntryError} from '../../utils/errors.js';

export const newComment = async (id, identityId, {content, reply_id}) => {
  try {
    const {rows} = await app.db.query(sql`
    INSERT 
      INTO comments (identity_id, post_id, content, reply_id)
      VALUES(${identityId}, ${id}, ${content}, ${reply_id})
      RETURNING *
    `);
    return rows[0];
  } catch (err) {
    throw new EntryError(err.message);
  }
};

export const updateComment = async (id, identityId, {content}) => {
  try {
    const {rows} = await app.db.query(sql`
    UPDATE comments SET
      content=${content}
    WHERE id=${id} AND identity_id=${identityId}
    RETURNING *
    `);
    return rows[0];
  } catch (err) {
    throw new EntryError(err.message);
  }
};

export const removeComment = async (id, identityId) => {
  return app.db.query(
    sql`DELETE FROM comments WHERE id=${id} AND identity_id=${identityId}`,
  );
};

export const comments = async (
  id,
  currentIdentity,
  {offset = 0, limit = 10},
) => {
  const {rows} = await app.db.query(sql`
  SELECT 
  COUNT(*) OVER () as total_count,
  c.*, i.type  as identity_type, i.meta as identity_meta,
  COALESCE(r.id IS NOT NULL, false) AS reported,
  EXISTS (
    SELECT id FROM likes WHERE post_id=${id} AND 
    identity_id=${currentIdentity} AND
    comment_id=c.id
  ) AS liked
  FROM comments c 
  JOIN identities i ON c.identity_id=i.id
  LEFT JOIN reports r ON r.comment_id=c.id AND r.identity_id=${currentIdentity}
  WHERE post_id=${id} AND reply_id IS NULL AND (r.blocked IS NULL OR r.blocked = false)
  ORDER BY created_at DESC  LIMIT ${limit} OFFSET ${offset}
  `);
  return rows;
};

export const commentsReplies = async (
  id,
  currentIdentity,
  {offset = 0, limit = 10},
) => {
  const {rows} = await app.db.query(sql`
  SELECT COUNT(*) OVER () as total_count,
  c.*, i.type  as identity_type, i.meta as identity_meta,
  COALESCE(r.id IS NOT NULL, false) AS reported,
  EXISTS (
    SELECT id FROM likes WHERE post_id=${id} AND 
    identity_id=${currentIdentity} AND
    comment_id=c.id
  ) AS liked
  FROM comments c 
  JOIN identities i ON c.identity_id=i.id
  LEFT JOIN reports r ON r.comment_id=c.id AND r.identity_id=${currentIdentity}
  WHERE reply_id=${id} AND (r.blocked IS NULL OR r.blocked = false)
  ORDER BY created_at DESC  LIMIT ${limit} OFFSET ${offset}
  `);
  return rows;
};

export const getComment = async (id) => {
  return app.db.get(sql`SELECT * FROM comments WHERE id=${id}`);
};
