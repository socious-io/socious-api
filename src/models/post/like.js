import sql from 'sql-template-tag'
import { app } from '../../index.js'
import { EntryError } from '../../utils/errors.js'

export const like = async (id, identityId, commentId = null) => {
  try {
    const { rows } = await app.db.query(sql`
    INSERT 
      INTO likes (identity_id, post_id, comment_id)
      VALUES(${identityId}, ${id}, ${commentId})
      RETURNING *
    `)
    return rows[0]
  } catch (err) {
    throw new EntryError(err.message)
  }
}

export const unlike = async (id, identityId, commentId = null) => {
  if (commentId !== null) {
    return app.db.query(
      sql`DELETE FROM likes WHERE post_id=${id} AND identity_id=${identityId} AND comment_id=${commentId}`
    )
  }
  return app.db.query(sql`DELETE FROM likes WHERE post_id=${id} AND identity_id=${identityId} AND comment_id IS NULL`)
}
