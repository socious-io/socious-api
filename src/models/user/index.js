import * as write from './write.js'
import * as read from './read.js'
import * as enums from './enums.js'
import * as additionals from './additionals.js'

export default {
  ...enums,
  ...read,
  ...write,
  ...additionals
}
