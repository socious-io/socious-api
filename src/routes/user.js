import Router from '@koa/router'
import User from '../models/user/index.js'
import Applicant from '../models/applicant/index.js'
import Auth from '../services/auth/index.js'
import Mission from '../models/mission/index.js'
import Offer from '../models/offer/index.js'
import Skill from '../models/skill/index.js'
import ImpactPoints from '../services/impact_points/index.js'
import Payment from '../services/payments/index.js'
import { paginate } from '../utils/middlewares/requests.js'
import { loginOptional, loginRequired } from '../utils/middlewares/authorization.js'
import Data, { validate } from '@socious/data'
import { checkIdParams } from '../utils/middlewares/route.js'
import { putContact } from '../services/sendgrid/index.js'
import { BadRequestError } from '../utils/errors.js'
import { recommendUserByUser, recommendProjectByUser, recommendOrgByUser } from '../services/recommender/index.js'

export const router = new Router()

router.get('/:id/profile', loginOptional, checkIdParams, async (ctx) => {
  if (ctx.user.status === User.StatusType.INACTIVE) {
    ctx.body = await User.getProfileLimited(ctx.params.id)
    return
  }
  ctx.body = await User.getProfile(ctx.params.id, ctx.identity.id)
})

router.post('/:id/report', loginRequired, async (ctx) => {
  await validate.ReportSchema.validateAsync(ctx.request.body)
  await User.report({
    ...ctx.request.body,
    user_id: ctx.params.id,
    identity_id: ctx.identity.id
  })

  ctx.body = {
    message: 'success'
  }
})

router.get('/by-username/:username/profile', loginOptional, async (ctx) => {
  if (ctx.user.status === User.StatusType.INACTIVE) {
    ctx.body = await User.getProfileByUsernameLimited(ctx.params.username)
    return
  }
  ctx.body = await User.getProfileByUsername(ctx.params.username, ctx.identity.id)
})

router.get('/profile', loginRequired, async (ctx) => {
  ctx.body = await User.getProfile(ctx.user.id)
})

router.post('/update/wallet', loginRequired, async (ctx) => {
  const wallet = ctx.request.body.wallet_address
  if (!wallet) throw new BadRequestError('wallet address is required')
  await User.updateWalletAddress(ctx.user.id, wallet)
  ctx.body = await User.getProfile(ctx.user.id)
})

router.post('/update/profile', loginRequired, async (ctx) => {
  await validate.UpdateProfileSchema.validateAsync(ctx.request.body)

  const skills = await Skill.getAllByNames(ctx.request.body.skills)
  ctx.request.body.skills = skills.map((s) => s.name)

  ctx.body = await User.updateProfile(ctx.user.id, ctx.request.body)

  putContact({
    first_name: ctx.body.first_name,
    last_name: ctx.body.last_name,
    email: ctx.body.email
  })
})

router.post('/change-password', loginRequired, async (ctx) => {
  await Auth.changePassword(ctx.user, ctx.request.body)
  ctx.body = {
    message: 'success'
  }
})

router.post('/change-password-direct', loginRequired, async (ctx) => {
  await Auth.directChangePassword(ctx.user, ctx.request.body)

  ctx.body = {
    message: 'success'
  }
})

router.post('/delete', loginRequired, async (ctx) => {
  await User.remove(ctx.user, ctx.request.body.reason)
  ctx.body = {
    message: 'success'
  }
})

router.get('/applicants', loginRequired, checkIdParams, paginate, async (ctx) => {
  ctx.body = await Applicant.getByUserId(ctx.user.id, ctx.paginate)
})

router.get('/:id/missions', loginRequired, checkIdParams, paginate, async (ctx) => {
  ctx.paginate.filter.assignee_id = ctx.params.id
  ctx.body = await Mission.getAll(ctx.paginate)
})

router.get('/missions', loginRequired, paginate, async (ctx) => {
  ctx.paginate.filter.assignee_id = ctx.identity.id

  const missions = await Mission.getAll(ctx.paginate)

  ctx.body = missions.map((m) => {
    return {
      ...m,
      ...Payment.amounts({
        amount: m.offer.assignment_total,
        service: Data.PaymentService.STRIPE
          ? m.offer.payment_mode === Data.PaymentMode.FIAT
          : Data.PaymentService.CRYPTO,
        verified: m.assigner.meta.verified_impact
      })
    }
  })
})

router.get('/offers', loginRequired, paginate, async (ctx) => {
  ctx.body = await Offer.getAll(ctx.identity.id, ctx.paginate)
})

router.post('/languages', loginRequired, async (ctx) => {
  await validate.ProfileAddLanguageSchema.validateAsync(ctx.request.body)
  ctx.body = await User.addLanguage(ctx.user, ctx.request.body)
})

router.post('/languages/update/:id', loginRequired, checkIdParams, async (ctx) => {
  await validate.ProfileAddLanguageSchema.validateAsync(ctx.request.body)
  ctx.body = await User.editLanguage(ctx.params.id, ctx.user, ctx.request.body)
})

router.post('/languages/remove/:id', loginRequired, checkIdParams, async (ctx) => {
  await User.removeLanguage(ctx.params.id, ctx.user)
  ctx.body = { message: 'success' }
})

router.post('/experiences', loginRequired, async (ctx) => {
  await validate.ProfileExperienceSchema.validateAsync(ctx.request.body)
  ctx.body = await User.addExperience(ctx.user, ctx.request.body)
})

router.post('/experiences/update/:id', loginRequired, checkIdParams, async (ctx) => {
  await validate.ProfileExperienceSchema.validateAsync(ctx.request.body)
  ctx.body = await User.editExperience(ctx.params.id, ctx.user, ctx.request.body)
})

router.post('/experiences/remove/:id', loginRequired, checkIdParams, async (ctx) => {
  ctx.body = await User.removeExperience(ctx.params.id, ctx.user)
})

router.get('/badges', loginRequired, async (ctx) => {
  ctx.body = { badges: await ImpactPoints.badges(ctx.identity.id) }
})

router.get('/:id/badges', loginRequired, checkIdParams, async (ctx) => {
  ctx.body = { badges: await ImpactPoints.badges(ctx.params.id) }
})

router.get('/impact-points', loginRequired, paginate, async (ctx) => {
  ctx.body = await ImpactPoints.history(ctx.identity.id, ctx.paginate)
})

router.post('/open-to-work', loginRequired, async (ctx) => {
  ctx.body = {
    open_to_work: await User.openToWork(ctx.user.id)
  }
})

router.post('/open-to-volunteer', loginRequired, async (ctx) => {
  ctx.body = {
    open_to_volunteer: await User.openToVolunteer(ctx.user.id)
  }
})

router.get('/:username/recommend/jobs', loginOptional, async (ctx) => {
  ctx.body = await recommendProjectByUser(ctx.params.username)
})

router.get('/:username/recommend/users', loginOptional, async (ctx) => {
  ctx.body = await recommendUserByUser(ctx.params.username)
})

router.get('/:username/recommend/orgs', loginOptional, async (ctx) => {
  ctx.body = await recommendOrgByUser(ctx.params.username)
})
