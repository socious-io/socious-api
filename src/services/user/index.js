import * as write from './write.js';
import * as read from './read.js';
import * as auth from './auth.js';

export default {
  ...auth,
  ...read,
  ...write,
};
