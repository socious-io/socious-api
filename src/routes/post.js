import Router from '@koa/router'
import { validate } from '@socious/data'
import Post from '../models/post/index.js'
import Notif from '../models/notification/index.js'
import Event from '../services/events/index.js'
import { paginate } from '../utils/middlewares/requests.js'
import { loginOptional, loginRequired } from '../utils/middlewares/authorization.js'
import { checkIdParams } from '../utils/middlewares/route.js'
export const router = new Router()

router.get('/:id', loginOptional, checkIdParams, async (ctx) => {
  ctx.body = await Post.get(ctx.params.id, ctx.identity.id)
})

router.get('/', loginOptional, paginate, async (ctx) => {
  ctx.body = await Post.all(ctx.identity.id, ctx.paginate)
})

router.post('/', loginRequired, async (ctx) => {
  await validate.PostSchema.validateAsync(ctx.request.body)
  ctx.body = await Post.insert(ctx.identity.id, ctx.request.body)
})

router.post('/:id/report', loginRequired, async (ctx) => {
  await validate.ReportSchema.validateAsync(ctx.request.body)
  await Post.report({
    ...ctx.request.body,
    post_id: ctx.params.id,
    identity_id: ctx.identity.id
  })

  ctx.body = {
    message: 'success'
  }
})

router.post('/update/:id', loginRequired, checkIdParams, async (ctx) => {
  await validate.PostSchema.validateAsync(ctx.request.body)
  await Post.permissioned(ctx.identity.id, ctx.params.id)
  ctx.body = await Post.update(ctx.params.id, ctx.identity.id, ctx.request.body)
})

router.post('/remove/:id', loginRequired, checkIdParams, async (ctx) => {
  await Post.permissioned(ctx.identity.id, ctx.params.id)
  await Post.remove(ctx.params.id)
  ctx.body = {
    message: 'success'
  }
})

router.get('/:id/comments', loginOptional, paginate, checkIdParams, async (ctx) => {
  ctx.body = await Post.comments(ctx.params.id, ctx.identity.id, ctx.paginate)
})

router.get('/comments/:id', loginOptional, paginate, checkIdParams, async (ctx) => {
  ctx.body = await Post.commentsReplies(ctx.params.id, ctx.identity.id, ctx.paginate)
})

router.post('/remove/comments/:id', loginRequired, checkIdParams, async (ctx) => {
  await validate.UUID.validateAsync(ctx.params.id)
  await Post.removeComment(ctx.params.id, ctx.identity.id)
  ctx.body = {
    message: 'success'
  }
})

router.post('/:id/comments', loginRequired, checkIdParams, async (ctx) => {
  await validate.CommentSchema.validateAsync(ctx.request.body)

  ctx.body = await Post.newComment(ctx.params.id, ctx.identity.id, ctx.request.body)

  const post = await Post.miniGet(ctx.params.id)

  Event.push(Event.Types.NOTIFICATION, post.identity_id, {
    type: Notif.Types.COMMENT,
    refId: ctx.body.id,
    parentId: post.id,
    identity: ctx.identity
  })
})

router.post('/comments/:id/report', loginRequired, async (ctx) => {
  await validate.ReportSchema.validateAsync(ctx.request.body)
  await Post.reportComment({
    ...ctx.request.body,
    comment_id: ctx.params.id,
    identity_id: ctx.identity.id
  })

  ctx.body = {
    message: 'success'
  }
})

router.post('/update/comments/:id', loginRequired, checkIdParams, async (ctx) => {
  await validate.CommentSchema.validateAsync(ctx.request.body)
  ctx.body = await Post.updateComment(ctx.params.id, ctx.identity.id, ctx.request.body)
})

router.post('/:id/like', loginRequired, checkIdParams, async (ctx) => {
  ctx.body = await Post.like(ctx.params.id, ctx.identity.id)

  const post = await Post.miniGet(ctx.params.id)

  Event.push(Event.Types.NOTIFICATION, post.identity_id, {
    type: Notif.Types.POST_LIKE,
    refId: ctx.body.id,
    parentId: post.id,
    identity: ctx.identity
  })
})

router.post('/:id/unlike', loginRequired, checkIdParams, async (ctx) => {
  await Post.unlike(ctx.params.id, ctx.identity.id)
  ctx.body = {
    message: 'success'
  }
})

router.post('/:id/comments/:comment_id/like', loginRequired, checkIdParams, async (ctx) => {
  ctx.body = await Post.like(ctx.params.id, ctx.identity.id, ctx.params.comment_id)

  const comment = await Post.getComment(ctx.params.comment_id)

  Event.push(Event.Types.NOTIFICATION, comment.identity_id, {
    type: Notif.Types.COMMENT_LIKE,
    refId: ctx.body.id,
    parentId: comment.post_id,
    identity: ctx.identity
  })
})

router.post('/:id/comments/:comment_id/unlike', loginRequired, checkIdParams, async (ctx) => {
  await Post.unlike(ctx.params.id, ctx.identity.id, ctx.params.comment_id)
  ctx.body = {
    message: 'success'
  }
})

router.post('/:id/share', loginRequired, checkIdParams, async (ctx) => {
  ctx.body = await Post.share(ctx.params.id, ctx.identity.id, ctx.request.body)
})
