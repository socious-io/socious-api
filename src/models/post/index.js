import * as write from './write.js'
import * as read from './read.js'
import * as comment from './comment.js'
import * as like from './like.js'

export default {
  ...read,
  ...write,
  ...comment,
  ...like
}
