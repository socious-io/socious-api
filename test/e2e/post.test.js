import supertest from 'supertest'
import { app } from '../../src/index.js'
import data from '../data/index.js'
import { registerAndVerify } from './globals/user.js'
import { create, getAll, getFiltered } from './globals/post.js'

let server, request

beforeAll(async () => {
  server = app.listen()
  request = supertest(server)

  await registerAndVerify(request, data)
})

test('create posts', async () => create(request, data))
test('get posts', async () => getAll(request, data))
test('filter posts', async () => getFiltered(request, data))

const cleanup = async () => {
  await app.db.query('DELETE FROM users')
  await app.db.pool.end()
}

afterAll((done) => {
  cleanup(done).then(() => {
    server.close(done)
  })
})
