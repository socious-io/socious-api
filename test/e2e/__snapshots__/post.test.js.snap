// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
{
  "error": ""email" must be a valid email",
}
`;

exports[`create posts 1`] = `
{
  "causes_tags": [
    "SOCIAL",
    "HEALTH",
  ],
  "content": "test content",
  "created_at": Any<String>,
  "deleted_at": null,
  "hashtags": [
    "TEST",
  ],
  "id": Any<String>,
  "identity_id": Any<String>,
  "identity_meta": Any<Object>,
  "identity_tags": null,
  "identity_type": "users",
  "liked": false,
  "likes": 0,
  "media": null,
  "reported": false,
  "search_tsv": "'content':2 'health':5 'social':4 'test':1,3",
  "shared": 0,
  "shared_from_identity": null,
  "shared_id": null,
  "shared_post": null,
  "updated_at": Any<String>,
}
`;

exports[`create posts 2`] = `
{
  "causes_tags": [
    "SOCIAL",
  ],
  "content": "test content 2",
  "created_at": Any<String>,
  "deleted_at": null,
  "hashtags": [
    "TEST2",
  ],
  "id": Any<String>,
  "identity_id": Any<String>,
  "identity_meta": Any<Object>,
  "identity_tags": null,
  "identity_type": "users",
  "liked": false,
  "likes": 0,
  "media": null,
  "reported": false,
  "search_tsv": "'2':3 'content':2 'social':5 'test':1 'test2':4",
  "shared": 0,
  "shared_from_identity": null,
  "shared_id": null,
  "shared_post": null,
  "updated_at": Any<String>,
}
`;

exports[`filter posts 1`] = `
{
  "items": [
    {
      "causes_tags": [
        "SOCIAL",
        "HEALTH",
      ],
      "content": "test content",
      "created_at": Any<String>,
      "deleted_at": null,
      "hashtags": [
        "TEST",
      ],
      "id": Any<String>,
      "identity_id": Any<String>,
      "identity_meta": Any<Object>,
      "identity_tags": null,
      "identity_type": "users",
      "liked": false,
      "likes": 0,
      "media": null,
      "reported": false,
      "search_tsv": "'content':2 'health':5 'social':4 'test':1,3",
      "shared": 0,
      "shared_from_identity": null,
      "shared_id": null,
      "shared_post": null,
      "updated_at": Any<String>,
    },
  ],
  "limit": 10,
  "page": 1,
  "total_count": 1,
}
`;

exports[`get posts 1`] = `
{
  "items": [
    {
      "causes_tags": [
        "SOCIAL",
      ],
      "content": "test content 2",
      "created_at": Any<String>,
      "deleted_at": null,
      "hashtags": [
        "TEST2",
      ],
      "id": Any<String>,
      "identity_id": Any<String>,
      "identity_meta": Any<Object>,
      "identity_tags": null,
      "identity_type": "users",
      "liked": false,
      "likes": 0,
      "media": null,
      "reported": false,
      "search_tsv": "'2':3 'content':2 'social':5 'test':1 'test2':4",
      "shared": 0,
      "shared_from_identity": null,
      "shared_id": null,
      "shared_post": null,
      "updated_at": Any<String>,
    },
    {
      "causes_tags": [
        "SOCIAL",
        "HEALTH",
      ],
      "content": "test content",
      "created_at": Any<String>,
      "deleted_at": null,
      "hashtags": [
        "TEST",
      ],
      "id": Any<String>,
      "identity_id": Any<String>,
      "identity_meta": Any<Object>,
      "identity_tags": null,
      "identity_type": "users",
      "liked": false,
      "likes": 0,
      "media": null,
      "reported": false,
      "search_tsv": "'content':2 'health':5 'social':4 'test':1,3",
      "shared": 0,
      "shared_from_identity": null,
      "shared_id": null,
      "shared_post": null,
      "updated_at": Any<String>,
    },
  ],
  "limit": 10,
  "page": 1,
  "total_count": 2,
}
`;
